user/tests/integration/application/src/request_handler.cpp

```diff
@@ -257,6 +259,38 @@ int RequestHandler::getLog(Request* req) {
     return 0;
 }
 
+int RequestHandler::readMailbox(Request* req) {
+    const auto runner = TestRunner::instance();
+    auto m = runner->peekOutboundMailbox();
+    if (!m) {
+        return 0;
+    }
+
+    CHECK(req->reply([&](JSONWriter& w) {
+        w.beginObject();
+        w.name("t").value((int)m->type());
+        const auto data = m->data();
+        if (data.size() > 0) {
+            w.name("d").value(data.data(), data.size());
+        }
+        w.endObject();
+    }));
+    if (m->waitedOn()) {
+        req->done(0, [](int, void* data) -> void {
```

Comment:
Why do we explicitly call `req->done()` here rather than let it be called in `RequestHandler::request(ctrl_request* ctrlReq)` like all the other request messages?

Reply:
It's not ideal, but just to keep the logic contained within `readMailbox()` and pass our own completion handler to `system_ctrl_set_result()`. If the mailbox entry is being actively waited on, we only want to unblock the waiter once the host has received it.

The second req->done() call in `request()` will just be no-op.